Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp

```diff
@@ -266,13 +266,40 @@ bool CanvasRenderingContext2DBase::isAccelerated() const
 void CanvasRenderingContext2DBase::reset()
 {
     unwindStateStack();
+
+    clearCanvas();
+
     m_stateStack.resize(1);
     m_stateStack.first() = State();
+
     m_path.clear();
     m_unrealizedSaveCount = 0;
     m_cachedImageData = std::nullopt;
 
     m_recordingContext = nullptr;
+
+    // Reset all drawing states to default value.
+    resetTransform();
+    resetClip();
+    setFillStyle({ Color::black });
+    setStrokeStyle({ Color::black });
+    setGlobalAlpha(1);
+    setGlobalCompositeOperation("source-over"_s);
+    setLineWidth(1);
+    setLineCap(CanvasLineCap::Butt);
+    setLineJoin(CanvasLineJoin::Miter);
+    setMiterLimit(10);
+    setLineDash({ });
+    setLineDashOffset(0);
+    setShadowOffsetX(0);
+    setShadowOffsetY(0);
+    setShadowBlur(0);
+    setShadowColor("rgba(0, 0, 0, 0)"_s);
+    setTextAlign(CanvasTextAlign::Start);
+    setTextBaseline(CanvasTextBaseline::Alphabetic);
+    setDirection(Direction::Inherit);
+    setImageSmoothingEnabled(true);
+    setImageSmoothingQuality(defaultSmoothingQuality);
```

Comment:
There's actually one bug here, which is that I must manually set all states to their default value using the `set...` methods before clearing the state stack then pushing a default one.

We need to manually set all these because clearing the state stack then pushing a default one is not enough. Since the canvas is backed by a graphics context, whenever the top state entry changes, we must also synchronize the state to the graphics context. For example, without the manual set, this code would still draw the rectangle at 0.5 alpha because when we reset, the alpha of the graphics context is not set back to 1:

```javascript
var canvas = document.getElementById("can");
var ctx = canvas.getContext("2d");

ctx.globalAlpha = 0.5;

ctx.reset();

ctx.fillStyle = "#00ff00";
ctx.fillRect(0, 0, 100, 100);
```

Reply:
We have two stacks here

1. CanvasRenderingContext2DBase::m_stateStack;
2. GraphicsContext::m_stack;

I think the design is to keep these two stacks synced all the times (except when m_unrealizedSaveCount > 0). This function flushes CanvasRenderingContext2DBase::m_stateStack and pushes the default State instead. But it does not touch the associated GraphicsContext::m_stack. I do not think this is right.

Also having to call these setting functions does not seem right and error prone. The resetting operation has to be atomic otherwise it will be easy to get it wrong at in the future. 

I think we need to make a deep clean-up here and recreate the ImageBuffer of the canvas instead of trying to reset its state to the default.

Reply:
> This function flushes CanvasRenderingContext2DBase::m_stateStack and pushes the default State instead. But it does not touch the associated GraphicsContext::m_stack. I do not think this is right.

Yeah this is what I meant.

> I think we need to make a deep clean-up here and recreate the ImageBuffer of the canvas instead of trying to reset its state to the default.

Sounds about right, I think I was going with this approach initially then changed my mind.

Offtopic, but when we create the ImageBuffer (and the associated graphics context), where do we apply the default state to the graphics context? Or it just so happens that the default graphics context state is the same as the default canvas state.

Reply:
I think `CanvasBase::setImageBuffer()` fixes the `GraphicsContext` of the backing store `ImageBuffer` to match the default `CanvasRenderingContext2DBase::State`.

```
    if (m_imageBuffer) {
        m_imageBuffer->context().setShadowsIgnoreTransforms(true);
        m_imageBuffer->context().setImageInterpolationQuality(defaultInterpolationQuality);
        m_imageBuffer->context().setStrokeThickness(1);
        ...
    }
```

Reply:
I was wrong. `CanvasRenderingContext2DBase::reset()` calls `unwindStateStack()` which flushes the state stack of the `GraphicsContext`. But still this does not restore it to the initial state.

I think also recreating the `ImageBuffer` might be too aggressive and might lead to memory complications since we need to free and allocate the same size of memory. So I would suggest implementing `GraphicsContext::reset()`.

```
void GraphicsContext::reset()
{
    // Unwind the context's state stack.
    while (context().stackSize())
        context().restore();

    ASSERT(m_stack.isEmpty());
    ASSERT(!m_transparencyLayerCount);

    auto defaultState = GraphicsContextState();
    mergeAllChanges(defaultState);
}


void CanvasRenderingContext2DBase::reset()
{
    m_stateStack.resize(1);
    m_stateStack.first() = State();
    m_path.clear();
    m_unrealizedSaveCount = 0;
    m_cachedImageData = std::nullopt;

    m_recordingContext = nullptr;

    if (auto* context = canvasBase().existingDrawingContext()) {
        context->reset();

        // Make state of context match the State of m_stateStack.
        context->setShadowsIgnoreTransforms(true);
        context->setImageInterpolationQuality(defaultInterpolationQuality);
        context->setStrokeThickness(1);
    }
}
```